hash.go

```diff
@@ -17,61 +17,86 @@ type HashToPoint interface {
 	Hash(in []byte) C.Point
 	// GetCurve returns the destination elliptic curve.
 	GetCurve() C.EllCurve
+	// GetHashToScalar returns a hash function that hashes strings to field elements.
+	GetHashToScalar() HashToScalar
 }
 
-type encoding struct {
-	E       C.EllCurve
-	Exp     Expander
-	Mapping M.MapToCurve
-	L       uint
+type HashToScalar interface {
+	// Hash returns an element of a field given a byte string.
+	Hash(in []byte) GF.Elt
+}
+
+type fieldEncoding struct {
+	F   GF.Field
+	Exp Expander
+	L   uint
+}
+
+// Hash deterministically hashes a string msg of any length into
+// an element of the given finite field.
+func (f *fieldEncoding) Hash(msg []byte) GF.Elt {
+	return f.hashToField(msg, 1)[0]
 }
 
 // hashToField is a function that hashes a string msg of any length into an
 // element of a finite field.
-func (e *encoding) hashToField(
+func (f *fieldEncoding) hashToField(
 	msg []byte, // msg is the message to hash.
-	count uint, // count is 1 or 2.
+	count uint, // count is 1 or 2 (the length of the result array).
 ) []GF.Elt {
-	F := e.E.Field()
-	m := F.Ext()
-	length := count * m * e.L
+	m := f.F.Ext()
+	length := count * m * f.L
 
-	pseudo := e.Exp.Expand(msg, length)
+	pseudo := f.Exp.Expand(msg, length)
 	u := make([]GF.Elt, count)
 	v := make([]interface{}, m)
-	p := F.P()
+	p := f.F.P()
 	for i := uint(0); i < count; i++ {
 		for j := uint(0); j < m; j++ {
-			offset := e.L * (j + i*m)
-			t := pseudo[offset : offset+e.L]
+			offset := f.L * (j + i*m)
+			t := pseudo[offset : offset+f.L]
 			vj := new(big.Int).SetBytes(t)
 			v[j] = vj.Mod(vj, p)
 		}
-		u[i] = F.Elt(v)
+		u[i] = f.F.Elt(v)
 	}
 	return u
 }
 
+type encoding struct {
+	E             C.EllCurve
+	Mapping       M.MapToCurve
+	FieldEncoding *fieldEncoding
+}
+
 func (e *encoding) GetCurve() C.EllCurve { return e.E }
 
 type encodeToCurve struct{ *encoding }
 
 func (s *encodeToCurve) IsRandomOracle() bool { return false }
 func (s *encodeToCurve) Hash(in []byte) C.Point {
-	u := s.hashToField(in, 1)
+	u := s.FieldEncoding.hashToField(in, 1)
 	Q := s.Mapping.Map(u[0])
 	P := s.E.ClearCofactor(Q)
 	return P
 }
 
+func (s *encodeToCurve) GetHashToScalar() HashToScalar {
+	return s.FieldEncoding
```

Comment:
However, the object returned must not be the field of the curve, but the field of the scalars.
You need to define the fields using the order of the curve as the prime modulus.